### PR TITLE
Redis support

### PR DESCRIPTION
**Pull request by danawillow**
Adds a new Redis resource and enables it in Terraform.

-----------------------------------------------------------------
# [all]
## [terraform]
Redis resource

### CODE REVIEW

File: templates/terraform/transport.go [hunk 1]
```diff
@@ -105,7 +106,16 @@ func sendRequest(config *Config, method, url string, body map[string]interface{}
...
 
-	req, err := http.NewRequest(method, url+"?alt=json", &buf)
+	u, err := urllib.Parse(url)
+	if err != nil {
+		return nil, err
+	}
+
+	queries := u.Query()
+	queries.Add("alt", "json")
+	newUrl := u.Scheme + "://" + u.Host + u.Path + "?" + queries.Encode()
```
Comment (danawillow) [inline templates/terraform/transport.go]: note to self: https://github.com/GoogleCloudPlatform/magic-modules/pull/191/files does it better

File: products/redis/api.yaml [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    exclude: true
+    base_url: |
+      projects/{{project}}/locations/{{region}}/instances?instanceId={{name}}
+    self_link: projects/{{project}}/locations/{{region}}/instances/{{name}}
+    description: |
+      A Google Cloud Redis instance.
+    input: true
+<%= indent(compile_file({}, 'templates/regional_async.yaml.erb'), 4) %>
+    parameters:
+      - !ruby/object:Api::Type::String # TODO: resourceref?
```
Comment (rosbo) [inline products/redis/api.yaml]: Why not?

File: products/redis/api.yaml [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+objects:
+  - !ruby/object:Api::Resource
+    name: 'Instance'
+    exclude: true
+    base_url: |
+      projects/{{project}}/locations/{{region}}/instances?instanceId={{name}}
+    self_link: projects/{{project}}/locations/{{region}}/instances/{{name}}
+    description: |
+      A Google Cloud Redis instance.
+    input: true
```
Comment (rosbo) [inline products/redis/api.yaml]: Add a reference block with the links to API documentation and guide:
```yaml
references: !ruby/object:Api::Resource::ReferenceLinks
      guides:
        'Reserving a Static External IP Address':
          'https://cloud.google.com/compute/docs/ip-addresses/reserve-static-external-ip-address'
      api: 'https://cloud.google.com/compute/docs/reference/latest/globalAddresses'
```

File: products/redis/api.yaml [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        name: currentLocationId
+        description: |
+          The current zone where the Redis endpoint is placed.
+          For Basic Tier instances, this will always be the same as the
+          [locationId] provided by the user at creation time. For Standard Tier
+          instances, this can be either [locationId] or [alternativeLocationId]
+          and can change after a failover event.
+        output: true
+      - !ruby/object:Api::Type::String
+        name: displayName
```
Comment (rosbo) [inline products/redis/api.yaml]: No update method for this one?

File: products/redis/terraform.yaml [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        custom_expand: 'templates/terraform/redis/expand_name.erb'
+        custom_flatten: 'templates/terraform/redis/flatten_name.erb'
+      redisVersion: !ruby/object:Provider::Terraform::PropertyOverride
+        default: !ruby/object:Provider::Terraform::Default
+          from_api: true
+      region: !ruby/object:Provider::Terraform::PropertyOverride
+        required: false
+        default: !ruby/object:Provider::Terraform::Default
+          from_api: true
+      reservedIpRange: !ruby/object:Provider::Terraform::PropertyOverride
```
Comment (rosbo) [inline products/redis/terraform.yaml]: Do we want to see :BASIC as the default value for `tier`? And set required to false

File: products/redis/api.yaml [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+          block, for example, 10.0.0.0/29 or 192.168.0.0/29. Ranges must be
+          unique and non-overlapping with existing subnets in an authorized
+          network.
+      - !ruby/object:Api::Type::Enum
+        name: tier
+        description: The service tier of the instance. #TODO describe values
+        values:
+          - :BASIC
+          - :STANDARD_HA
+        required: true
```
Comment (nelsonjr) [inline products/redis/api.yaml]: new line at end?

File: products/redis/api.yaml [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+    name: beta
+    base_url: https://redis.googleapis.com/v1beta1/
+scopes:
+   # TODO this is fake and made up but MM requires it
+  - https://www.googleapis.com/auth/redis
+objects:
+  - !ruby/object:Api::Resource
+    name: 'Instance'
+    exclude: true
+    base_url: |
```
Comment (nelsonjr) [inline products/redis/api.yaml]: A few things here:

1. The `base_url` is to provide a collection. You probably want `self_link_query` instead
2. Keep `base_url: projects/{{project}}/locations/{{region}}/instances`
3. Remove `self_link`

Comment (danawillow) [reply products/redis/api.yaml]: Can you explain what a collection is in this context?

The reason I did it this way is that in the Terraform templates, we use `base_url` (via `collection_url`) to create the resource, and then `self_link` (via `self_link_url`) to read, update, and delete it. For redis, the query params are in the create call, so we would call POST to projects/{{project}}/locations/{{region}}/instances?instanceId={{name}} to create it. Then once we want to read it though, it's a GET to projects/{{project}}/locations/{{region}}/instances/{{name}}. The way it is now outputs the correct URLs for the Terraform templates- would you change it to use custom code, or is this fine?

**Comment by danawillow** [comment]
Looking into rspec failures.

**Comment by ndmckinley** [comment]
I'm filling in for a robot who works on MagicModules PRs!

The robot tried to run tests, but:

```
+ go test -v ./google -parallel 16 -run '^Test' -timeout 1m
# github.com/terraform-providers/terraform-provider-google/google
google/transport.go:79:37: undefined: raw
google/transport.go:108:12: undefined: urllib
FAIL	github.com/terraform-providers/terraform-provider-google/google [build failed]
```

Unfortunately, the robot's not smart enough yet to complain when things don't work.  That's on the roadmap!

File: templates/terraform/resource.html.markdown.erb [hunk 8]
```diff
@@ -91,6 +91,7 @@ In addition to the arguments listed above, the following computed attributes are
 <%= lines(build_nested_property_documentation(prop)) -%>
 <% end -%>
 
+<% unless object.async.nil? -%>
```
Comment (ndmckinley) [inline templates/terraform/resource.html.markdown.erb]: Shoot, I forgot to do that on mine.  Good catch.

**Comment by danawillow** [comment]
All right, the generated code in Terraform is looking good from my perspective, so this is ready for real reviews.

**Pull request merged**